File: apps/merchant-portal/src/lib/api/addresses.ts
import { createApiClient, ApiResponse } from "./client";

export interface PickupAddress {
  id: string;
  user_id: string;
  tenant_id: string;
  label: string;
  address: string;
  city: string;
  province: string;
  postal_code: string;
  country: string;
  lat: number | null;
  lng: number | null;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreatePickupAddressPayload {
  label: string;
  address: string;
  city?: string;
  province?: string;
  postal_code?: string;
  country?: string;
  lat?: number;
  lng?: number;
}

export const addressesApi = {
  async list(): Promise<PickupAddress[]> {
    const client = createApiClient();
    const res = await client.get<ApiResponse<PickupAddress[]>>("/v1/pickup-addresses");
    return res.data.data;
  },

  async create(payload: CreatePickupAddressPayload): Promise<PickupAddress> {
    const client = createApiClient();
    const res = await client.post<ApiResponse<PickupAddress>>("/v1/pickup-addresses", payload);
    return res.data.data;
  },

  async setDefault(id: string): Promise<void> {
    const client = createApiClient();
    await client.patch(`/v1/pickup-addresses/${id}/default`);
  },

  async remove(id: string): Promise<void> {
    const client = createApiClient();
    await client.delete(`/v1/pickup-addresses/${id}`);
  },
};

File: services/identity/src/api/http/addresses.rs
use axum::{extract::{Path, State}, Json};
use serde::Deserialize;
use std::sync::Arc;
use uuid::Uuid;
use logisticos_auth::middleware::AuthClaims;
use logisticos_errors::AppError;
use crate::{api::http::AppState, domain::entities::PickupAddress};

#[derive(Debug, Deserialize)]
pub struct CreatePickupAddressBody {
    pub label:       String,
    pub address:     String,
    pub city:        Option<String>,
    pub province:    Option<String>,
    pub postal_code: Option<String>,
    pub country:     Option<String>,
    pub lat:         Option<f64>,
    pub lng:         Option<f64>,
}

pub async fn list(
    AuthClaims(claims): AuthClaims,
    State(state): State<Arc<AppState>>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = logisticos_types::UserId::from_uuid(claims.user_id);
    let addrs = state.address_repo.list_by_user(&user_id).await
        .map_err(AppError::Internal)?;
    Ok(Json(serde_json::json!({ "data": addrs })))
}

pub async fn create(
    AuthClaims(claims): AuthClaims,
    State(state): State<Arc<AppState>>,
    Json(body): Json<CreatePickupAddressBody>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id   = logisticos_types::UserId::from_uuid(claims.user_id);
    let tenant_id = logisticos_types::TenantId::from_uuid(claims.tenant_id);

    let addr = PickupAddress::new(
        user_id,
        tenant_id,
        body.label,
        body.address,
        body.city.unwrap_or_default(),
        body.province.unwrap_or_default(),
        body.postal_code.unwrap_or_default(),
        body.country.unwrap_or_else(|| "PH".to_string()),
        body.lat,
        body.lng,
    );
    state.address_repo.insert(&addr).await.map_err(AppError::Internal)?;
    Ok(Json(serde_json::json!({ "data": addr })))
}

pub async fn delete(
    AuthClaims(claims): AuthClaims,
    Path(id): Path<Uuid>,
    State(state): State<Arc<AppState>>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = logisticos_types::UserId::from_uuid(claims.user_id);
    state.address_repo.delete(id, &user_id).await.map_err(AppError::Internal)?;
    Ok(Json(serde_json::json!({ "ok": true })))
}

pub async fn set_default(
    AuthClaims(claims): AuthClaims,
    Path(id): Path<Uuid>,
    State(state): State<Arc<AppState>>,
) -> Result<Json<serde_json::Value>, AppError> {
    let user_id = logisticos_types::UserId::from_uuid(claims.user_id);
    state.address_repo.set_default(id, &user_id).await.map_err(AppError::Internal)?;
    Ok(Json(serde_json::json!({ "ok": true })))
}
